src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,851 @@
+        }
+
+        let mut tags = BTreeMap::new();
+        tags.insert("endpoint".into(), config_to_endpoint(&config));
+        tags.insert(
+            "host".into(),
+            match &config.get_hosts().get(0).expect("host is not missed") {
+                Host::Tcp(host) => host.clone(),
+                #[cfg(unix)]
+                Host::Unix(path) => path.to_string_lossy().to_string(),
+            },
+        );
+
+        let mut this = Self {
+            config,
+            tls_config,
+            client: None,
+            namespace,
+            tags,
+            datname_filter,
+        };
+        this.build_client().await?;
+        Ok(this)
+    }
+
+    async fn build_client(&mut self) -> Result<(), BuildError> {
+        let client = match &self.tls_config {
+            Some(tls_config) => {
+                let mut builder =
+                    SslConnector::builder(SslMethod::tls_client()).context(TlsFailed)?;
+                builder
+                    .set_ca_file(tls_config.ca_file.clone())
+                    .context(TlsFailed)?;
+                let connector = MakeTlsConnector::new(builder.build());
+
+                let (client, connection) =
+                    self.config
+                        .connect(connector)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+            None => {
+                let (client, connection) =
+                    self.config
+                        .connect(NoTls)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+        };
+
+        let version_row = client
+            .query_one("SELECT version()", &[])
+            .await
+            .with_context(|| VersionFailed {
+                endpoint: config_to_endpoint(&self.config),
+            })?;
+        let version = version_row
+            .try_get::<&str, &str>("version")
+            .with_context(|| VersionFailed {
+                endpoint: config_to_endpoint(&self.config),
+            })?;
+        debug!(message = "Connected to server.", endpoint = %config_to_endpoint(&self.config), server_version = %version);
+
+        self.client = Some(client);
+        Ok(())
+    }
+
+    async fn collect(&mut self) -> stream::BoxStream<'static, Metric> {
+        let build_client = match self.client {
+            Some(_) => Ok(()),
+            None => self.build_client().await,
+        };
+
+        let metrics = match build_client {
+            Ok(()) => self
+                .collect_metrics(self.client.as_ref().expect("should exists at this point"))
```

Comment:
I still feel weird about structuring this this way where it is required to have called `self.build_client()` already. What would you think of something like:

```rust
 // get_client() would handle checking if the client already exists or making a new one if not
let metrics = match self.get_client().await { // Result<Client, _>
            Ok(client) => // do scrape and return those metrics plus an up metric of 1
            Err(err) => // emit error and return an up metric of 0
}
```

where `get_client` is something like:

```rust
async fn get_client(&self) -> Result<Client, _> {
    match self.client {
            Some(_) => Ok(client),
            None => self.build_client().await,
        };
} 
```

Reply:
The problem with this approach a lifetime.
`build_client` require `&mut self` because of change `self.client` from `None` to `Some`
this means that `get_client` also will require `&mut self`
once we take `client` from `get_client` which required `&mut self` we can not call `collect_metrics` while result of `get_client` is not dropped
It's a little wired and I was surprised myself when tried the approach described above.

Reply:
Ah, yeah, I see what you are saying.

I think one approach would be to split up some of the logic rather than putting it all on `self` (`PostgresqlMetrics`). My suggestion would be to wrap the `tokio_postgres::Client` with our own type that is able to reconnect on query failure. It seems like we only call `query()` so this shouldn't be too much additional implementation. Then we could avoid wrapping the `Client` in an `Option` just to drive that reconnect behavior.

An alternative would be to move `collect_metrics` out from `PostgresqlMetrics` so that we don't need to borrow it again. This would require passing through the dependencies of those functions.

It's just that something seems a bit off with the way it is currently structured in that we have implicit assumptions.